crates/xmtp-networking/src/grpc_api_helper.rs

```diff
@@ -53,44 +127,76 @@ pub async fn query_serialized(
 pub async fn publish(
     host: String,
     token: String,
-    json_envelopes: String,
+    envelopes: Vec<v1::Envelope>,
 ) -> Result<v1::PublishResponse, tonic::Status> {
     let host = host.to_string();
-    let channel = Channel::from_shared(host)
-        .map_err(|e| tonic::Status::new(tonic::Code::Internal, format!("{}", e)))?
-        .connect()
-        .await
-        .map_err(|e| tonic::Status::new(tonic::Code::Internal, format!("{}", e)))?;
-
+    // Prepare the auth token as a MetadataValue
     let auth_token_string = format!("Bearer {}", token);
     let token: MetadataValue<_> = auth_token_string
         .parse()
         .map_err(|e| tonic::Status::new(tonic::Code::Internal, format!("{}", e)))?;
 
-    let mut client = v1::message_api_client::MessageApiClient::with_interceptor(
-        channel,
-        move |mut req: Request<()>| {
-            req.metadata_mut().insert("authorization", token.clone());
-            Ok(req)
-        },
-    );
-
-    let mut request = v1::PublishRequest::default();
-    // Deserialize the JSON string into a vector of Envelopes
-    let envelopes: Vec<v1::Envelope> = serde_json::from_str(&json_envelopes)
-        .map_err(|e| tonic::Status::new(tonic::Code::Internal, format!("{}", e)))?;
-    request.envelopes = envelopes;
-    let response = client.publish(request).await;
+    let request = v1::PublishRequest { envelopes };
+
+    // TODO: See above about DRY-ing up clients. Probably want to re-use the same client
+    // in the future.
+    let response = if host.starts_with("https://") {
+        println!("Using TLS client");
+        // Set up the TLS client
+        let tls = tls_config()
+            .map_err(|e| tonic::Status::new(tonic::Code::Internal, format!("{}", e)))?;
+        let mut http = HttpConnector::new();
+        http.enforce_http(false);
+        let uri = Uri::from_str(&host)
+            .map_err(|e| tonic::Status::new(tonic::Code::Internal, format!("{}", e)))?;
+        let connector = tower::ServiceBuilder::new()
+            .layer_fn(move |s| {
+                let tls = tls.clone();
+
+                hyper_rustls::HttpsConnectorBuilder::new()
+                    .with_tls_config(tls)
+                    .https_or_http()
+                    .enable_http2()
+                    .wrap_connector(s)
+            })
+            .service(http);
+        let client = hyper::Client::builder().build(connector);
```

Comment:
I see what you mean about the need to DRY this up

Reply:
yeah :/ my naive attempt at just taking this code and pasting it into a function resulted in some indecipherable Rust errors (at least with my knowledge at this point)